go/WorkQueue.go

```diff
@@ -120,6 +122,77 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
 	return err
 }
 
+// **************-This method can be used only when database is locked-**************
+func (workQueue *WorkQueue) AddAtomicItem(ctx context.Context, db *redis.Client, item Item) error {
+	txf := func(tx *redis.Tx) error {
+		pipe := db.Pipeline()
+
+		mainQueueLenCmd := pipe.LLen(ctx, workQueue.mainQueueKey)
+		processingQueueLenCmd := pipe.LLen(ctx, workQueue.processingKey)
+
+		_, err := pipe.Exec(ctx)
+
+		mainQueueLen, err := mainQueueLenCmd.Result()
+
+		processingQueueLen, err := processingQueueLenCmd.Result()
+
+		pipeNew := db.Pipeline()
+		processingItemsInQueue, err := pipeNew.LPos(ctx, workQueue.processingKey, item.ID, redis.LPosArgs{
+			Rank:   1,
+			MaxLen: processingQueueLen,
+		}).Result()
+
+		workingItemsInQueue, err := pipeNew.LPos(ctx, workQueue.mainQueueKey, item.ID, redis.LPosArgs{
+			Rank:   1,
+			MaxLen: mainQueueLen,
+		}).Result()
+
+		pipeNew.Exec(ctx)
+
+		if workingItemsInQueue == 0 || processingItemsInQueue == 0 {
+			return nil
+		}
+
+		pipeEx := tx.Pipeline()
+		workQueue.AddItemToPipeline(ctx, pipeEx, item)
+		_, err = pipeEx.Exec(ctx)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	err := db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
+	//fmt.Println(err)
+	for err == redis.TxFailedErr {
+		if err == nil {
+			// Success.
+			return nil
+		} else if err == redis.TxFailedErr {
+			err = db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
+		}
+	}
+	return nil
+}
+
+func (workQueue *WorkQueue) GetQueueLengths(ctx context.Context, db *redis.Client) ([]int64, error) {
```

Comment:
There's no need to return a slice here! It's much more idiomatic to return two ints.

For example:

```go
func (workQueue *WorkQueue) Lengths(ctx context.Context, db *redis.Client) (int64, int64, error)
```

You can also annotate the return values to make it more readable:

```go
func (workQueue *WorkQueue) Lengths(ctx context.Context, db *redis.Client) (queueLen, processing int64, error)
```

Reply:
Also, if you ever are returning a fixed length slice, return an array instead (i.e. `[2]int64`)